include: "rules/common.smk"

rule all:
    input:
        generate_plots,
        expand([
                "results/{analysis_id}/1_rename/{sample}.fasta",
                "results/{analysis_id}/2_six_frames/1_screening/{sample}.domtxt",
                "results/{analysis_id}/2_six_frames/profiles_hits.txt",
                "results/{analysis_id}/4_ORF/coding_summary.txt",
                "results/{analysis_id}/4_ORF/3_functional_annot_tables/.finished",
                "results/{analysis_id}/5_phylogenies/markers_classification.txt",
                "results/{analysis_id}/6_protein_clustering/shared_content_taxonomy.txt",
                "results/{analysis_id}/7_ANI/ani_genus_species_taxonomy.txt",

                "results/{analysis_id}/aggregated_signals.txt",
                "results/{analysis_id}/crassus_results.tsv",

                #"results/{analysis_id}/prefinal_table.txt"
                ],
                sample=SAMPLES, analysis_id=ANALYSES_IDS)



include: "rules/assembly.smk"
include: "rules/scan_contigs.smk"
include: "rules/open_reading_frames.smk"
include: "rules/phylogenetic_trees.smk"
include: "rules/protein_clustering.smk"
include: "rules/assess_taxonomy.smk"
include: "rules/genomes_alignments.smk"
